theano/tensor/basic.py

```diff
@@ -3965,18 +3970,27 @@ def _make_node_internal(self, axis, tensors,
     def perform(self, node, axis_and_tensors, out_):
         out, = out_
         axis, tensors = axis_and_tensors[0], axis_and_tensors[1:]
-        ndim = tensors[0].ndim
-        if axis < -ndim:
-            raise IndexError("Join axis %d out of bounds [0, %d)" %
-                             (axis, ndim))
-
-        out[0] = theano._asarray(numpy.concatenate(tensors, axis=axis),
-                                 dtype=node.outputs[0].type.dtype)
+        nonempty_tensors = [0 if tensor.size == 0 else 1 for tensor in tensors]
+        num_nonempty = numpy.sum(nonempty_tensors)
```

Comment:
with the code change above, you can do any(num_nonempty), this would be epsilon faster.